File: verify.js
#!/usr/bin/env node

/**
 * Verification script for The Stylist frontend application.
 * Checks that all required components are properly set up.
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// Add colors for terminal output
const GREEN = '\x1b[32m';
const YELLOW = '\x1b[33m';
const RED = '\x1b[31m';
const BLUE = '\x1b[34m';
const RESET = '\x1b[0m';

// Define required files and directories
const requiredFiles = [
  'package.json',
  'tsconfig.json',
  'webpack.config.js',
  'src/StylistWidget.tsx',
  'src/components/ChatWidget/ChatWidget.tsx',
  'src/components/VirtualTryOn/VirtualTryOn.tsx',
  'src/hooks/useTryOn.ts',
  'src/hooks/useImageProcessing.ts',
  'src/services/background-removal/removeBackgroundApi.ts',
  'src/services/background-removal/tfBackgroundRemoval.ts',
  'public/models/segmentation-model/model.json',
];

// Define required dependencies
const requiredDependencies = [
  'react',
  'react-dom',
  'zustand',
  'axios',
];

// Define required dev dependencies
const requiredDevDependencies = [
  'typescript',
  'webpack',
  'babel',
  'jest',
];

// Check if a file exists
function checkFileExists(filePath) {
  try {
    return fs.existsSync(filePath);
  } catch (err) {
    return false;
  }
}

// Check required files
function checkRequiredFiles() {
  console.log(`${BLUE}Checking required files...${RESET}`);
  
  const missingFiles = [];
  
  for (const file of requiredFiles) {
    if (!checkFileExists(file)) {
      missingFiles.push(file);
    }
  }
  
  if (missingFiles.length > 0) {
    console.warn(`${YELLOW}Missing files: ${missingFiles.join(', ')}${RESET}`);
    return false;
  }
  
  console.log(`${GREEN}✓ All required files are present${RESET}`);
  return true;
}

// Check package.json
function checkPackageJson() {
  console.log(`${BLUE}Checking package.json...${RESET}`);
  
  try {
    const packageJson = JSON.parse(fs.readFileSync('package.json', 'utf-8'));
    
    // Check for required scripts
    const requiredScripts = ['start', 'build', 'test'];
    const missingScripts = [];
    
    for (const script of requiredScripts) {
      if (!packageJson.scripts || !packageJson.scripts[script]) {
        missingScripts.push(script);
      }
    }
    
    if (missingScripts.length > 0) {
      console.warn(`${YELLOW}Missing scripts in package.json: ${missingScripts.join(', ')}${RESET}`);
      return false;
    }
    
    // Check for required dependencies
    const missingDependencies = [];
    
    for (const dep of requiredDependencies) {
      if (!packageJson.dependencies || !Object.keys(packageJson.dependencies).some(d => d === dep || d.startsWith(`${dep}@`))) {
        missingDependencies.push(dep);
      }
    }
    
    if (missingDependencies.length > 0) {
      console.warn(`${YELLOW}Missing dependencies in package.json: ${missingDependencies.join(', ')}${RESET}`);
      return false;
    }
    
    // Check for required dev dependencies
    const missingDevDependencies = [];
    
    for (const dep of requiredDevDependencies) {
      if (!packageJson.devDependencies || !Object.keys(packageJson.devDependencies).some(d => d === dep || d.startsWith(`${dep}`) || d.includes(dep))) {
        missingDevDependencies.push(dep);
      }
    }
    
    if (missingDevDependencies.length > 0) {
      console.warn(`${YELLOW}Missing dev dependencies in package.json: ${missingDevDependencies.join(', ')}${RESET}`);
      return false;
    }
    
    console.log(`${GREEN}✓ package.json is properly configured${RESET}`);
    return true;
  } catch (err) {
    console.error(`${RED}Error reading package.json: ${err.message}${RESET}`);
    return false;
  }
}

// Check webpack config
function checkWebpackConfig() {
  console.log(`${BLUE}Checking webpack configuration...${RESET}`);
  
  try {
    const webpackConfig = fs.readFileSync('webpack.config.js', 'utf-8');
    
    // Check for common webpack configurations
    const requiredConfigs = [
      'entry', 
      'output', 
      'module', 
      'resolve',
      '.tsx?',  // TypeScript loader
      'scss',   // SCSS loader
    ];
    
    const missingConfigs = [];
    
    for (const config of requiredConfigs) {
      if (!webpackConfig.includes(config)) {
        missingConfigs.push(config);
      }
    }
    
    if (missingConfigs.length > 0) {
      console.warn(`${YELLOW}Missing webpack configurations: ${missingConfigs.join(', ')}${RESET}`);
      return false;
    }
    
    console.log(`${GREEN}✓ webpack configuration is properly set up${RESET}`);
    return true;
  } catch (err) {
    console.error(`${RED}Error reading webpack.config.js: ${err.message}${RESET}`);
    return false;
  }
}

// Check TypeScript config
function checkTypeScriptConfig() {
  console.log(`${BLUE}Checking TypeScript configuration...${RESET}`);
  
  try {
    const tsConfig = JSON.parse(fs.readFileSync('tsconfig.json', 'utf-8'));
    
    // Check for required compiler options
    const requiredOptions = ['jsx', 'target', 'module', 'strict'];
    const missingOptions = [];
    
    for (const option of requiredOptions) {
      if (!tsConfig.compilerOptions || !tsConfig.compilerOptions[option]) {
        missingOptions.push(option);
      }
    }
    
    if (missingOptions.length > 0) {
      console.warn(`${YELLOW}Missing TypeScript compiler options: ${missingOptions.join(', ')}${RESET}`);
      return false;
    }
    
    console.log(`${GREEN}✓ TypeScript configuration is properly set up${RESET}`);
    return true;
  } catch (err) {
    console.error(`${RED}Error reading tsconfig.json: ${err.message}${RESET}`);
    return false;
  }
}

// Check TensorFlow.js model files
function checkTensorFlowModel() {
  console.log(`${BLUE}Checking TensorFlow.js model files...${RESET}`);
  
  try {
    const modelDir = 'public/models/segmentation-model';
    const modelJson = path.join(modelDir, 'model.json');
    
    if (!checkFileExists(modelJson)) {
      console.warn(`${YELLOW}Missing TensorFlow.js model file: ${modelJson}${RESET}`);
      console.info(`${BLUE}See instructions in ${modelDir}/README.md${RESET}`);
      return false;
    }
    
    try {
      const modelConfig = JSON.parse(fs.readFileSync(modelJson, 'utf-8'));
      
      // Check if the model.json has the required fields
      if (!modelConfig.weightsManifest) {
        console.warn(`${YELLOW}Invalid model.json: missing weightsManifest${RESET}`);
        console.info(`${BLUE}See instructions in ${modelDir}/README.md${RESET}`);
        return false;
      }
      
      // Check if weights file exists
      const weightsFiles = modelConfig.weightsManifest[0]?.paths || [];
      if (weightsFiles.length === 0) {
        console.warn(`${YELLOW}No weights files specified in model.json${RESET}`);
        console.info(`${BLUE}See instructions in ${modelDir}/README.md${RESET}`);
        return false;
      }
      
    } catch (err) {
      console.warn(`${YELLOW}Invalid model.json: ${err.message}${RESET}`);
      console.info(`${BLUE}See instructions in ${modelDir}/README.md${RESET}`);
      return false;
    }
    
    console.log(`${GREEN}✓ TensorFlow.js model files are present${RESET}`);
    return true;
  } catch (err) {
    console.error(`${RED}Error checking TensorFlow.js model files: ${err.message}${RESET}`);
    return false;
  }
}

// Main verification function
function verify() {
  console.log(`${BLUE}Verifying The Stylist frontend setup...${RESET}`);
  
  const checks = [
    checkRequiredFiles,
    checkPackageJson,
    checkWebpackConfig,
    checkTypeScriptConfig,
    checkTensorFlowModel,
  ];
  
  let success = true;
  
  for (const check of checks) {
    if (!check()) {
      success = false;
    }
    console.log(''); // Add a line break between checks
  }
  
  if (success) {
    console.log(`${GREEN}========================================${RESET}`);
    console.log(`${GREEN}       PROJECT IS READY FOR TESTING     ${RESET}`);
    console.log(`${GREEN}========================================${RESET}`);
    return 0;
  } else {
    console.warn(`${YELLOW}========================================${RESET}`);
    console.warn(`${YELLOW}    PROJECT SETUP NEEDS ATTENTION    ${RESET}`);
    console.warn(`${YELLOW}========================================${RESET}`);
    console.info(`${BLUE}Please address the warnings above and run this script again.${RESET}`);
    return 1;
  }
}

// Run the verification
process.exit(verify());

File: verify.py
#!/usr/bin/env python3
"""
Verification script for The Stylist backend application.
Checks that all required components are properly set up.
"""

import os
import sys
import importlib.util
import logging
import json
from pathlib import Path
from dotenv import load_dotenv

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s",
)
logger = logging.getLogger(__name__)

# Add colors for terminal output
GREEN = "\033[92m"
YELLOW = "\033[93m"
RED = "\033[91m"
BLUE = "\033[94m"
ENDC = "\033[0m"

def check_env_variables():
    """Check environment variables."""
    load_dotenv()
    
    required_vars = [
        "REMOVE_BG_API_KEY",
        "ANTHROPIC_API_KEY",
        "STYLIST_API_KEY",
    ]
    
    missing_vars = []
    
    for var in required_vars:
        if not os.getenv(var):
            missing_vars.append(var)
    
    if missing_vars:
        logger.warning(f"{YELLOW}Missing environment variables: {', '.join(missing_vars)}{ENDC}")
        logger.info(f"{BLUE}Please check stillremaining.txt for details on required variables{ENDC}")
        return False
    
    logger.info(f"{GREEN}✓ Environment variables verified{ENDC}")
    return True

def check_python_dependencies():
    """Check that required Python dependencies are installed."""
    required_modules = [
        "fastapi",
        "uvicorn",
        "pydantic",
        "requests",
        "aiohttp",
        "redis",
        "PIL",
        "tensorflow",
        "jsonschema",
    ]
    
    missing_modules = []
    
    for module in required_modules:
        if not importlib.util.find_spec(module):
            if module == "PIL":
                # Pillow is imported as PIL
                try:
                    __import__("PIL")
                except ImportError:
                    missing_modules.append("Pillow")
            else:
                missing_modules.append(module)
    
    if missing_modules:
        logger.warning(f"{YELLOW}Missing Python dependencies: {', '.join(missing_modules)}{ENDC}")
        logger.info(f"{BLUE}Run `pip install -r requirements.txt` to install missing dependencies{ENDC}")
        return False
    
    logger.info(f"{GREEN}✓ Python dependencies verified{ENDC}")
    return True

def check_tensorflow_model():
    """Check if TensorFlow.js model files exist."""
    base_dir = Path("public/models/segmentation-model")
    
    required_files = [
        "model.json",
    ]
    
    missing_files = []
    
    for file in required_files:
        if not (base_dir / file).exists():
            missing_files.append(str(base_dir / file))
    
    if missing_files:
        logger.warning(f"{YELLOW}Missing TensorFlow.js model files: {', '.join(missing_files)}{ENDC}")
        logger.info(f"{BLUE}See instructions in public/models/segmentation-model/README.md{ENDC}")
        return False
    
    logger.info(f"{GREEN}✓ TensorFlow.js model files verified{ENDC}")
    return True

def check_api_routes():
    """Check that API routes are properly defined."""
    api_files = [
        "api/recommendation_routes.py",
        "api/retailer_routes.py",
        "api/inventory_routes.py",
    ]
    
    missing_files = []
    
    for file in api_files:
        if not Path(file).exists():
            missing_files.append(file)
    
    if missing_files:
        logger.warning(f"{YELLOW}Missing API route files: {', '.join(missing_files)}{ENDC}")
        return False
    
    logger.info(f"{GREEN}✓ API routes verified{ENDC}")
    return True

def check_frontend_files():
    """Check that frontend files are properly defined."""
    required_files = [
        "src/StylistWidget.tsx",
        "src/components/ChatWidget/ChatWidget.tsx",
        "src/components/VirtualTryOn/VirtualTryOn.tsx",
        "src/hooks/useTryOn.ts",
        "package.json",
    ]
    
    missing_files = []
    
    for file in required_files:
        if not Path(file).exists():
            missing_files.append(file)
    
    if missing_files:
        logger.warning(f"{YELLOW}Missing frontend files: {', '.join(missing_files)}{ENDC}")
        return False
    
    # Check package.json
    try:
        with open("package.json", "r") as f:
            package_data = json.load(f)
        
        required_scripts = ["start", "build", "test"]
        missing_scripts = []
        
        for script in required_scripts:
            if script not in package_data.get("scripts", {}):
                missing_scripts.append(script)
        
        if missing_scripts:
            logger.warning(f"{YELLOW}Missing scripts in package.json: {', '.join(missing_scripts)}{ENDC}")
            return False
        
    except Exception as e:
        logger.warning(f"{YELLOW}Error reading package.json: {str(e)}{ENDC}")
        return False
    
    logger.info(f"{GREEN}✓ Frontend files verified{ENDC}")
    return True

def check_docker_setup():
    """Check Docker setup."""
    if not Path("Dockerfile").exists():
        logger.warning(f"{YELLOW}Dockerfile not found{ENDC}")
        return False
    
    if not Path("nginx.conf").exists():
        logger.warning(f"{YELLOW}nginx.conf not found{ENDC}")
        return False
    
    logger.info(f"{GREEN}✓ Docker setup verified{ENDC}")
    return True

def check_github_actions():
    """Check GitHub Actions setup."""
    if not Path(".github/workflows/ci-cd.yml").exists():
        logger.warning(f"{YELLOW}.github/workflows/ci-cd.yml not found{ENDC}")
        return False
    
    logger.info(f"{GREEN}✓ GitHub Actions setup verified{ENDC}")
    return True

def main():
    """Run all verification checks."""
    logger.info(f"{BLUE}Verifying The Stylist installation...{ENDC}")
    
    checks = [
        check_env_variables,
        check_python_dependencies,
        check_tensorflow_model,
        check_api_routes,
        check_frontend_files,
        check_docker_setup,
        check_github_actions,
    ]
    
    success = True
    
    for check in checks:
        if not check():
            success = False
    
    if success:
        logger.info(f"\n{GREEN}========================================{ENDC}")
        logger.info(f"{GREEN}       PROJECT IS READY FOR TESTING     {ENDC}")
        logger.info(f"{GREEN}========================================{ENDC}")
        return 0
    else:
        logger.warning(f"\n{YELLOW}========================================{ENDC}")
        logger.warning(f"{YELLOW}    PROJECT SETUP NEEDS ATTENTION    {ENDC}")
        logger.warning(f"{YELLOW}========================================{ENDC}")
        logger.info(f"{BLUE}Please address the warnings above and run this script again.{ENDC}")
        return 1

if __name__ == "__main__":
    sys.exit(main())